src/Workspaces/Remote/ServiceHub/Services/LanguageServer.cs

```diff
@@ -0,0 +1,155 @@
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    // we need per language server for now since ILanguageClient
+    // doesn't allow multiple content types to be associated with
+    // one language server
+    internal class CSharpLanguageServer : LanguageServer
+    {
+        public CSharpLanguageServer(Stream stream, IServiceProvider serviceProvider)
+            : base(stream, serviceProvider, LanguageNames.CSharp)
+        {
+        }
+    }
+
+    internal class VisualBasicLanguageServer : LanguageServer
+    {
+        public VisualBasicLanguageServer(Stream stream, IServiceProvider serviceProvider)
+            : base(stream, serviceProvider, LanguageNames.VisualBasic)
+        {
+        }
+    }
+
+    internal abstract class LanguageServer : ServiceBase
+    {
+        private static readonly IImmutableSet<string> s_supportedKinds =
+            ImmutableHashSet.Create(
+                NavigateToItemKind.Class,
+                NavigateToItemKind.Constant,
+                NavigateToItemKind.Delegate,
+                NavigateToItemKind.Enum,
+                NavigateToItemKind.EnumItem,
+                NavigateToItemKind.Event,
+                NavigateToItemKind.Field,
+                NavigateToItemKind.Interface,
+                NavigateToItemKind.Method,
+                NavigateToItemKind.Module,
+                NavigateToItemKind.Property,
+                NavigateToItemKind.Structure);
+
+        private readonly string _languageName;
+
+        public LanguageServer(Stream stream, IServiceProvider serviceProvider, string languageName)
+            : base(serviceProvider, stream, SpecializedCollections.EmptyEnumerable<JsonConverter>())
+        {
+            _languageName = languageName;
+
+            StartService();
+        }
+
+        [JsonRpcMethod(Methods.InitializeName)]
+        public object Initialize(int? processId, string rootPath, Uri rootUri, ClientCapabilities capabilities, TraceSetting trace, CancellationToken cancellationToken)
+        {
+            // our LSP server only supports WorkspaceStreamingSymbolProvider capability
+            // for now
+            return new InitializeResult()
+            {
+                Capabilities = new VSServerCapabilities()
+                {
+                    WorkspaceStreamingSymbolProvider = true
+                }
+            };
+        }
+
+        [JsonRpcMethod(Methods.InitializedName)]
+        public Task Initialized()
+        {
+            return Task.CompletedTask;
+        }
+
+        [JsonRpcMethod(Methods.ShutdownName)]
+        public void Shutdown(CancellationToken cancellationToken)
+        {
+            // our language server shutdown when VS shutdown
+        }
+
+        [JsonRpcMethod(VSSymbolMethods.WorkspaceBeginSymbolName)]
+        public Task<VSBeginSymbolParams> WorkspaceBeginSymbolAsync(string query, int searchId, CancellationToken cancellationToken)
```

Comment:
this protocol is temporary. so let's leave it like it for now.

what I want at the end is having LSP Server in LSP layer you created. and here. I just get stream and connect that stream to LSP server in LSP layer.

basically, in 
public CSharpLanguageServer(Stream stream, IServiceProvider serviceProvider)
        {
        }

we would do
var rpc = new JsonRpc(Stream, new LanguageServer());

rather than what I am doing here.

Reply:
If this is temporary it should have TODO with explanation and link to an issue.

---
In reply to: [306227540](https://github.com/dotnet/roslyn/pull/37390#discussion_r306227540) [](ancestors = 306227540)